### PR TITLE
Update profiling CLI arguments

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

PR adds/updates CLI flags:
- `--profiling-events-log-path <file>` enable the event logging with ability to specify the file to store the messages between the runtime and the language server
- `--profiling-path <file>` enable the application profiling with ability to specify the file to store the `.npss` sampling statistics
- `--profiling-time <seconds>` limits the application profiling time. If not specified, the profiling is stopped on the application shutdown

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [ ] ~Unit tests have been written where possible.~
  - [ ] ~If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.~

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala [hunk 1]
```diff
@@ -220,6 +234,17 @@ class MainModule(serverConfig: LanguageServerConfig, logLevel: LogLevel) {
       "capability-router"
     )
 
+  val methodsSampler =
+    serverConfig.profilingConfig.profilingPath match {
+      case Some(path) =>
+        val s = new TempFileSampler(path.toFile)
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala]: I guess we no longer need to name the class `TempFileSampler`.

File: engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistry.scala [hunk 2]
```diff
@@ -110,11 +113,11 @@ final class ContextRegistry(
           .foreach(_ ! update)
 
       case update: Api.ExecutionFailed =>
-        sampler.stop(6.seconds)(context.dispatcher)
+        sampler.stop(profilingTimeout)(context.dispatcher)
         store.getListener(update.contextId).foreach(_ ! update)
 
       case update: Api.ExecutionComplete =>
-        sampler.stop(6.seconds)(context.dispatcher)
+        sampler.stop(profilingTimeout)(context.dispatcher)
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistry.scala]: The sampler is started at the beginning and it runs until the timeout is reached. Then the file is written to disk and sampling stops. Is that what we want, @kazcw , @MichaelMauderer ? If the IDE specifies small timeout, recording may stop sooner than user performs the action. The IDE can specify huge timeout, but then: when the file is going to be saved to disk?

Comment (kazcw) [reply engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistry.scala]: If profiling is triggered only by timeouts that would make batch mode slow and/or flaky, depending on the timeouts used.

In the frontend we usually save the profile just before exit. When the frontend exits, it causes the backend to shut down too. Can the backend save the profile before exiting?

Comment (JaroslavTulach) [reply engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistry.scala]: Following could work:
- when no timeout is specified, it defaults to infinity
- when the engine language server receives "exit request", it `stop`s profiling and dumps the sampler data

We may need further tweaks in the future, but for now this shall be good enough for the _batch mode_ purposes. I think.

**Comment by JaroslavTulach** [comment]
One more request. The IDE logs events it sends to the language server engine runtime. Can we make sure the same events, with the same IDs and similar timestamps are in the generated `.log` file? We need this to pair the two logs - having additional events in the log isn't necessary - in fact too many events may complicate orientation on the timeline.

File: engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistry.scala (hunk 2 above)
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistry.scala]: I updated the logic

Comment (MichaelMauderer) [reply engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistry.scala]: If this is implemented, please also update the PR description.

**Review by kazcw — APPROVED** [review]

**Review by MichaelMauderer — APPROVED** [review]

File: engine/language-server/src/main/scala/org/enso/languageserver/boot/LanguageServerComponent.scala [hunk 3]
```diff
@@ -62,21 +67,41 @@ class LanguageServerComponent(config: LanguageServerConfig, logLevel: LogLevel)
...
   /** @inheritdoc */
   override def stop(): Future[ComponentStopped.type] =
     maybeServerCtx match {
       case None =>
-        Future.failed(new Exception("Server isn't running"))
+        Future.successful(ComponentStopped)
 
       case Some(serverContext) =>
         for {
+          _ <- stopSampling(serverContext)
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/boot/LanguageServerComponent.scala:92]: Yup, this looks like fine order of actions. Stopping sampling before terminating Truffle.

**Review by JaroslavTulach — APPROVED** [review]
Let's get this in and let's see how well it works!

**Comment by 4e6** [comment]
> Can we make sure the same events, with the same IDs and similar timestamps are in the generated .log file?

The logging goes through the central logging service. I updated the logic, and not it also produces the profiling log when the profiling is enabled. This way the log captures all the messages as the previous solution with the Java log handler was only able to capture logs from within the runtime.

**Pull request merged**